packages/@aws-cdk/aws-appmesh/lib/virtual-router-listener.ts

```diff
@@ -0,0 +1,122 @@
+import * as cdk from '@aws-cdk/core';
+import { CfnVirtualRouter } from './appmesh.generated';
+import { Protocol } from './shared-interfaces';
+
+/**
+ * Represents the properties needed to define Listeners for a VirtualRouter
+ */
+export interface RouterListenerProps {
+  /**
+   * Port to listen for connections on
+   *
+   * @default - 8080
+   */
+  readonly port?: number
+}
+
+/**
+ * Properties for a VirtualRouter listener
+ */
+export interface VirtualRouterListenerConfig {
+  /**
+   * Single listener config for a VirtualRouter
+   */
+  readonly listener: CfnVirtualRouter.VirtualRouterListenerProperty;
+}
+
+/**
+ * Represents the properties needed to define listeners for a VirtualRouter
+ */
+export abstract class VirtualRouterListener {
+  /**
+   * Returns an HTTP Listener for a VirtualRouter
+   */
+  public static httpVirtualRouterListener(props: RouterListenerProps = {}): VirtualRouterListener {
+    return new HttpVirtualRouterListener(props);
+  }
+
+  /**
+   * Returns an HTTP2 Listener for a VirtualRouter
+   */
+  public static http2GatewayListener(props: RouterListenerProps = {}): VirtualRouterListener {
+    return new Http2VirtualRouterListener(props);
+  }
+
+  /**
+   * Returns a GRPC Listener for a VirtualRouter
+   */
+  public static grpcGatewayListener(props: RouterListenerProps = {}): VirtualRouterListener {
+    return new GrpcVirtualRouterListener(props);
+  }
+
+  /**
+   * Returns a TCP Listener for a VirtualRouter
+   */
+  public static tcpGatewayListener(props: RouterListenerProps = {}): VirtualRouterListener {
+    return new TcpVirtualRouterListener(props);
+  }
+
+  /**
+   * Protocol the listener implements
+   */
+  protected abstract protocol: Protocol;
+
+  /**
+   * Port to listen for connections on
+   */
+  protected abstract port: number;
+
+  /**
+   * Called when the VirtualRouterListener type is initialized. Can be used to enforce
+   * mutual exclusivity
+   */
+  public abstract bind(scope: cdk.Construct): VirtualRouterListenerConfig;
+}
+
+class HttpVirtualRouterListener extends VirtualRouterListener {
+  /**
+   * Protocol the listener implements
+   */
+  protected protocol: Protocol = Protocol.HTTP;
```

Comment:
Can you instead make the base class abstract with a `public abstract readonly protocol: Protocol` property, and then each of the subclasses will implement `protocol`?

Reply:
Yeah, we can further simplify this by only having one subclass for all 4 protocols.